students/jinee_han/lesson05/Mail3.py

```diff
@@ -0,0 +1,183 @@
+    :param donation_amount:
+    :return: nothing
+    '''
+    donor_db[donation_name].append(donation_amount)
+
+
+def donor_present(donor_name):
+    '''
+    Check if the donor name is present in the dictionary
+    :param donor_name:
+    :return: True if donor present, False if not
+    '''
+
+    # Use a comprehension to reduce code to check if donor is present
+    return len([name for name in donor_db.keys() if name == donor_name]) > 0
+
+
+def display_list():
+    '''
+    Display the list of current donors
+    :return: prints donor names
+    '''
+    for keys in donor_db:
+        print(keys)
+
+
+def add_donor(donor_name):
+    '''
+    Add donor name to the dictionary
+    :param donor_name:
+    :return: nothing
+    '''
+    donation_history = []
+    donor_db.update({donor_name:donation_history})
+
+
+def save_thank_you_to_file(donor_name, donation_amount, thank_you_note_formatted):
+    '''
+    Save the thank you note to file
+    :param donor_name: The current donor
+    :param donation_amount: The donation amount
+    :param thank_you_note_formatted: The formatted note to write to file
+    :return: nothing
+    '''
+    filename = "{}_{}_{}donations.txt".format(donor_name, donation_amount, len(donor_db[donor_name]))
+
+    with open("{}".format(filename), "w") as f:
+        f.write(thank_you_note_formatted)
+
+
+def format_thank_you_note(donator, donation_amount):
+    '''
+    Format the thank you note before writing to file
+    :param donator: The current donator
+    :param donation_amount: The donation amount
+    :return: nothing
+    '''
+    list_object = []
+    list_object.append("Dear {},\n\n".format(donator.title())  )
+    list_object.append("\tThank you for your very kind donation of  ${:10.2f}!\n".format(donation_amount)  )
+    list_object.append("\tIt will be put to very good use.\n\n")
+    list_object.append("\t\t\tSincerely,\n")
+    list_object.append("\t\t\t\t\t - The Team")
+
+    thank_you_note_formatted = " ".join(list_object)
+    save_thank_you_to_file(donator, donation_amount, thank_you_note_formatted)
+
+
+def create_report():
+    '''
+    Create a report of all donators, and amounts in sorted order with most donations first
+    :return: Prints values to console
+    '''
+    """Generate a tabular report of donation history"""
+    header ='\n{:^18}|{:^13}|{:^13}|{:>13}'.format("Donor Name", "Total Given", "Num Gifts", "Average Gift")
+    print(header)
+    print('-'*len(header))
+    sorted_list = sorted(donor_db.items(), key=lambda x: sum(x[1]), reverse = True)
+    for donor in sorted_list:
+        total = sum(donor[1])
+        num = len(donor[1])
+        average = total/num
+        print('{:^18} ${:>12,.2f}{:^13}  ${:>12,.2f}'.format(donor[0],total,num,average))
+    print('')
+
+
+def send_thank_you_to_all_donors():
+    '''
+    Send a thank you note to all donors
+    :return: nothing
+    '''
+    for donor in donor_db:
+        format_thank_you_note(donor, sum(donor_db[donor]))
```

Comment:
Note this is creating a letter for each donation for each donor - can you make a single letter for each donor?